### PR TITLE
Extend explanations and update signature of creating data frames in Westeros tutorials

### LINKED ISSUE #502: Remove deprecated signature of creating data frames in tutorials with `make_df`
Using the new signature to create the data frame with the `make_df()` function in the tutorials might increase the understanding of them.
As I understood correctly the old signature needs the following code to create the data frame (e.g. data frame from the [`westeros_emissions_bounds`](https://github.com/iiasa/message_ix/blob/v3.3.0/tutorial/westeros/westeros_emissions_bounds.ipynb) tutorial): 
```
base_emission_factor = {
    'node_loc': country,
    'year_vtg': vintage_years,
    'year_act': act_years,
    'mode': 'standard',
    'unit': 'tCO2/kWa',
} 

emission_factor = make_df(base_emission_factor, technology= 'coal_ppl', emission= 'CO2', value= 7.4)
```

Which can also be achieved by the new signature without the need to build a dictionary to append:
```
emission_factor = make_df("emission_factor", 
                           node_loc=country, 
                           year_vtg=vintage_years, 
                           year_act=act_years, 
                           mode='standard', 
                           unit='tCO2/kWa', 
                           technology= 'coal_ppl',
                           emission= 'CO2', 
                           value= 7.4 )
```

Any thoughts @OFR-IIASA and @khaeru? 
I'm happy to change it, if wanted.

### PR DESCRIPTION
**Pull request by LauWien**
This PR extends and updates the Westeros tutorials as well as extends the "Examples" section in the `make_df` documentation. PR #523 should be merged before merging this PR. After doing that the solution of the objective value of the Westeros baseline tutorial needs to be updated again to let the tutorial tests pass.
Closes #502. 

## Done in this PR
- Extended explanations incl. documentation for the utility function `make_df`
- Updated signature of creating data frames.
- Updated calculation of the `historical_new_capacity` and the corresponding tests 
- Applied code style `black`.
- Updated the quotation of variables and keywords.

## How to review

- Note that the CI checks all pass.
- Note the release notes were updated.
- Run `jupyter notebook` and go through the westeros baseline tutorial.
  - See if the explanations are correct/understandable, especially the historical activity and historical new capacity.
  - See if the quoting style of the different variables, keywords are readable as is.
  - See if new calculation of the `historical_new_capacity` is correct.
- Note the difference in the files of the `message_ix/testing` and `message_ix/tests` folder due to the updated calculation of `historical_new_capacity`. 
- Build the documentation and look at the [documentation for the utility function](https://docs.messageix.org/en/stable/api.html#module-message_ix.util) and the restructuring of the [tutorial overview page](https://docs.messageix.org/en/stable/tutorials.html#westeros-electrified).
- Possible, skim through the other tutorials.

## PR checklist

- [x] Continuous integration checks all ✅
- [x] Update release notes.

### CODE REVIEW

File: tutorial/westeros/westeros_baseline.ipynb [hunk 1]
```diff
@@ -492,19 +506,23 @@
...
+    "    \"year_act\": act_years,\n",
+    "    \"mode\": \"standard\",\n",
+    "    \"time\": \"year\",\n",
+    "    \"unit\": \"-\",\n",
     "}\n",
     "\n",
-    "base_input = make_df(base, node_origin=country, time_origin='year')\n",
-    "base_output = make_df(base, node_dest=country, time_dest='year')"
+    "# Extend the base dataframe specific for the \"input\" parameter\n",
+    "base_input = make_df(base, node_origin=country, time_origin=\"year\")\n",
```
Comment (khaeru) [inline tutorial/westeros/westeros_baseline.ipynb:522]: This is the deprecated signature, right? I'll try to update it to use the preferred one.

File: tutorial/westeros/westeros_soft_constraints.ipynb [hunk 2]
```diff
@@ -107,22 +106,25 @@
...
    "source": [
-    "df = pd.DataFrame({\n",
-    "    'node_loc': country,\n",
-    "    'technology': 'wind_ppl',\n",
-    "    'year_act': model_horizon,\n",
-    "    'time': 'year',\n",
-    "    'value': [.00, .0, .0, .01],\n",
-    "    'unit': '-'})\n",
-    "scen.add_par('soft_activity_up', df)"
+    "df = pd.DataFrame(\n",
```
Comment (khaeru) [inline tutorial/westeros/westeros_soft_constraints.ipynb:109]: In some cases like here, `black` was applied but the underlying code idiom was left as-is. The code works and it's fine to merge, but I think a later PR could improve these further to suggest a common style, e.g.:
```py
name = "soft_activity_up"
scen.add_par(
    name,
    make_df(
        name,
        node_loc=country,
        technology="wind_ppl",
        year_act=model_horizon,
        time="year",
        value=[0, 0, 0, 0.01],
        unit="-",
    )
)
```
or
```py
name = "soft_activity_up"
data = make_df(
    name,
    node_loc=country,
    technology="wind_ppl",
    year_act=model_horizon,
    time="year",
    value=[0, 0, 0, 0.01],
    unit="-",
)
scen.add_par(name, data)
```

File: tutorial/westeros/westeros_addon_technologies.ipynb [hunk 3]
```diff
@@ -231,16 +233,18 @@
...
-    "tec = 'hs_house'\n",
-    "scen.add_set('technology', tec)\n",
+    "tec = \"hs_house\"\n",
+    "scen.add_set(\"technology\", tec)\n",
     "\n",
-    "hs_house_out = make_df(base_output, technology=tec, commodity='heat', \n",
-    "                   level='useful', value=1.0)\n",
-    "scen.add_par('output', hs_house_out)\n",
+    "hs_house_out = make_df(\n",
+    "    base_output, technology=tec, commodity=\"heat\", level=\"useful\", value=1.0\n",
```
Comment (khaeru) [inline tutorial/westeros/westeros_addon_technologies.ipynb:240]: This is still the deprecated usage.

File: tutorial/westeros/westeros_firm_capacity.ipynb [hunk 4]
```diff
@@ -151,25 +157,25 @@
...
-    "# add the ratings as a set \n",
-    "scen.add_set('rating', ['r1', 'r2'])\n",
+    "# Add the ratings as a set\n",
+    "scen.add_set(\"rating\", [\"r1\", \"r2\"])\n",
     "\n",
-    "# for Wind PPL\n",
-    "rating_bin = make_df(base_reliability, technology= 'wind_ppl', value = 0.2, rating= 'r1')\n",
-    "scen.add_par('rating_bin', rating_bin)\n",
+    "# For the wind power plant\n",
+    "rating_bin = make_df(base_df, technology=\"wind_ppl\", value=0.2, rating=\"r1\")\n",
```
Comment (khaeru) [inline tutorial/westeros/westeros_firm_capacity.ipynb:164]: This is still the deprecated usage.

File: tutorial/westeros/westeros_flexible_generation.ipynb [hunk 5]
```diff
@@ -125,105 +118,113 @@
...
+    "        \"level\": \"secondary\",\n",
+    "        \"unit\": \"-\",\n",
+    "        \"time\": \"year\",\n",
+    "        \"year_act\": model_horizon,\n",
+    "    }\n",
+    ")\n",
     "\n",
-    "rating_bin = make_df(base_rating, technology= 'wind_ppl', value = 0.2, rating= 'r1')\n",
-    "scen.add_par('rating_bin', rating_bin)\n",
+    "rating_bin = make_df(base_rating, technology=\"wind_ppl\", value=0.2, rating=\"r1\")\n",
```
Comment (khaeru) [inline tutorial/westeros/westeros_flexible_generation.ipynb:154]: This is still the deprecated usage.

File: tutorial/westeros/westeros_historical_new_capacity.ipynb [hunk 6]
```diff
@@ -531,61 +539,61 @@
...
+    "    \"node_loc\": country,\n",
+    "    \"year_vtg\": vintage_years,\n",
+    "    \"year_act\": act_years,\n",
+    "    \"mode\": \"standard\",\n",
+    "    \"time\": \"year\",\n",
+    "    \"unit\": \"-\",\n",
     "}\n",
     "\n",
-    "base_output = make_df(base, node_dest=country, time_dest='year')\n",
+    "base_output = make_df(base, node_dest=country, time_dest=\"year\")\n",
```
Comment (khaeru) [inline tutorial/westeros/westeros_historical_new_capacity.ipynb:553]: This is still the deprecated usage.

File: tutorial/westeros/westeros_baseline.ipynb [hunk 7]
```diff
@@ -697,34 +741,21 @@
...
-   },
-   "outputs": [],
-   "source": [
-    "base_technical_lifetime = {\n",
-    "    'node_loc': country,\n",
-    "    'year_vtg': model_horizon,\n",
-    "    'unit': 'y',\n",
-    "}"
+    "| Value                                               | Condition                                           |\n",
+    "|:----------------------------------------------------|:----------------------------------------------------|\n",
```
Comment (khaeru) [inline tutorial/westeros/westeros_baseline.ipynb:751]: Here I noticed that the table looked strange, i.e. everything was right-aligned in my Jupyter/browser.

I found https://stackoverflow.com/questions/48655801/tables-in-markdown-in-jupyter which suggested using `|:--|` as a column specifier to explicitly force left alignment, whereas `|---|` is implicit and seems to be handled differently by Jupyter than other Markdown implementations.

There are many other tables in the tutorials, and I'm not asking that you revise all of these in order to approve the PR; just making a note for posterity.

**Review by khaeru — CHANGES_REQUESTED** [review]
I pushed some changes, but ended up running out of time (after ~2 hours) to make further ones.

The main change remaining: some files still use the deprecated signature for `make_df()` with another data frame as the first argument. As I understood, the goal of this PR was to remove these. In the commit “Remove deprecated make_df() signature in westeros_baseline tutorial” (currently 3ffb6e82bdd6d623af0b89bf14721227272a22de) I show the kind of changes that could be made to use the signature/idiom recommended by the documentation. I've left comments on just 1 instance (there are multiple) in each of the 3 remaining affected files.

**Comment by LauWien** [comment]
> I pushed some changes, but ended up running out of time (after ~2 hours) to make further ones.
> 
> The main change remaining: some files still use the deprecated signature for `make_df()` with another data frame as the first argument. As I understood, the goal of this PR was to remove these. In the commit “Remove deprecated make_df() signature in westeros_baseline tutorial” (currently [3ffb6e8](https://github.com/iiasa/message_ix/commit/3ffb6e82bdd6d623af0b89bf14721227272a22de)) I show the kind of changes that could be made to use the signature/idiom recommended by the documentation. I've left comments on just 1 instance (there are multiple) in each of the 3 remaining affected files.

Thank you very much for your review and the time you have spent on it!

File: tutorial/westeros/westeros_baseline.ipynb (hunk 7 above)
Comment (LauWien) [reply tutorial/westeros/westeros_baseline.ipynb:751]: I just applied it for all the tutorials, so we can put it out of our mind.

**Comment by LauWien** [comment]
I should've addressed now all your points. Many thanks again for your review and time :)

**Review by khaeru — APPROVED** [review]
Thanks for the revisions @LauWien—impressively thorough!

I'm not sure why the CLA check is complaining, but I'll merge this.

**Pull request merged**